### ISSUE TITLE
Failing Mocker tests in PHP7 and Unit::_message()/var_export() does not handle circular refs

### ISSUE THREAD

**Issue by alexbowers** [UnionOfRAD/lithium#1240]
ErrorException (E_WARNING) thrown.
 File    : /var/www/vhosts/lithium/test/Unit.php
 Class   : lithium\test\Unit
 Method  : lithium\test\{closure}()
 Line    : 1872
 ________
var_export does not handle circular references
 ________
var_export, line ??
lithium\test\Unit::_message(), line 1872
lithium\test\Unit::assert(), line 237
lithium\test\Unit::assertArrayHasKey(), line 957
lithium\tests\cases\test\MockerTest::testMagicCallGetStoredResultsWhenCalled(), line 416
lithium\test\Unit::lithium\test\{closure}() @ ?, line 1611
lithium\core\Object::_filter(), line 263
lithium\test\Unit::_runTestMethod(), line 1615
lithium\test\Unit::run(), line 213
lithium\util\Collection::invoke(), line 193
lithium\util\Collection::__call(), line 207
lithium\test\Report::run(), line 147
lithium\test\Dispatcher::run()::{closure} @ 65, line 69
lithium\core\StaticObject::_filter(), line 140
lithium\test\Dispatcher::run(), line 73
lithium\console\command\Test::run()::{closure} @ 235, line 236
lithium\console\command\Test::lithium\console\command\{closure}() @ ?, line 145
lithium\console\command\Test::run(), line 238
lithium\core\Object::invokeMethod(), line 179
lithium\console\Command::__invoke(), line 139
lithium\console\Dispatcher::lithium\console\{closure}() @ ?, line 198
lithium\core\StaticObject::_filter(), line 140
lithium\console\Dispatcher::_call(), line 201
lithium\core\StaticObject::invokeMethod(), line 84
lithium\console\Dispatcher::run()::{closure} @ 95, line 105
lithium\core\StaticObject::_filter(), line 140
lithium\console\Dispatcher::run(), line 109
/var/www/vhosts/lithium/console/lithium.php::{closure} @ 74, line 75
[main], line 110
 ________

ErrorException (E_WARNING) thrown.
 File    : /var/www/vhosts/lithium/test/Unit.php
 Class   : lithium\test\Unit
 Method  : lithium\test\{closure}()
 Line    : 1872
 ________
var_export does not handle circular references
 ________
var_export, line ??
lithium\test\Unit::_message(), line 1872
lithium\test\Unit::assert(), line 237
lithium\test\Unit::assertArrayHasKey(), line 957
lithium\tests\cases\test\MockerTest::testMagicCallGetStoredResultsWhenCalledIndirectly(), line 436
lithium\test\Unit::lithium\test\{closure}() @ ?, line 1611
lithium\core\Object::_filter(), line 263
lithium\test\Unit::_runTestMethod(), line 1615
lithium\test\Unit::run(), line 213
lithium\util\Collection::invoke(), line 193
lithium\util\Collection::__call(), line 207
lithium\test\Report::run(), line 147
lithium\test\Dispatcher::run()::{closure} @ 65, line 69
lithium\core\StaticObject::_filter(), line 140
lithium\test\Dispatcher::run(), line 73
lithium\console\command\Test::run()::{closure} @ 235, line 236
lithium\console\command\Test::lithium\console\command\{closure}() @ ?, line 145
lithium\console\command\Test::run(), line 238
lithium\core\Object::invokeMethod(), line 179
lithium\console\Command::__invoke(), line 139
lithium\console\Dispatcher::lithium\console\{closure}() @ ?, line 198
lithium\core\StaticObject::_filter(), line 140
lithium\console\Dispatcher::_call(), line 201
lithium\core\StaticObject::invokeMethod(), line 84
lithium\console\Dispatcher::run()::{closure} @ 95, line 105
lithium\core\StaticObject::_filter(), line 140
lithium\console\Dispatcher::run(), line 109
/var/www/vhosts/lithium/console/lithium.php::{closure} @ 74, line 75
[main], line 110
 ________

**Comment by alexbowers** [comment]
@davidpersson How come Li3 doesn't use an established unit testing tool like phpunit? 

The issue with this currently appears to be in the underlying testing framework.

**Comment by davidpersson** [comment]
Running the test under different versions should not yield different results. Note that the recursion triggers the error (we don't want any recursion). This is in: `lithium\tests\cases\test\MockerTest::testMagicCallGetStoredResultsWhenCalled(), line 41`

Under PHP 5.6:
```
  ["__call"]=>
  array(1) {
    [0]=>
    array(3) {
      ["args"]=>
      array(2) {
        [0]=>
        string(3) "foo"
        [1]=>
        array(0) {
        }
      }
      ["result"]=>
      bool(true)
      ["time"]=>
      float(1451758013.1694)
    }
  }
  ["methodFoo"]=>
  array(1) {
    [0]=>
    array(3) {
      ["args"]=>
      array(0) {
      }
      ["result"]=>
      bool(true)
      ["time"]=>
      float(1451758013.1693)
    }
  }
}
```

Under PHP 7.0.3-dev:
```
  ["__call"]=>
  array(1) {
    [0]=>
    array(3) {
      ["args"]=>
      array(2) {
        [0]=>
        array(2) {
          [0]=>
          object(lithium\tests\mocks\test\mockStdClass\MockDelegate)#49 (5) {
            ["parent"]=>
            object(lithium\tests\mocks\test\mockStdClass\Mock)#47 (4) {
              ["mocker"]=>
              *RECURSION*
              ["results"]=>
              array(1) {
                ["__call"]=>
                *RECURSION*
              }
              ["_safeVars":protected]=>
              array(6) {
                [0]=>
                string(8) "_classes"
                [1]=>
                string(14) "_methodFilters"
                [2]=>
                string(6) "mocker"
                [3]=>
                string(9) "_safeVars"
                [4]=>
                string(7) "results"
                [5]=>
                string(13) "staticResults"
              }
              ["_methodFilters":protected]=>
              array(0) {
              }
            }
            ["_data":protected]=>
            array(0) {
            }
            ["_config":protected]=>
            array(1) {
              ["init"]=>
              bool(true)
            }
            ["_autoConfig":protected]=>
            array(0) {
            }
            ["_methodFilters":protected]=>
            array(0) {
            }
          }
          [1]=>
          string(6) "__call"
        }
        [1]=>
        array(0) {
        }
      }
      ["result"]=>
      bool(true)
      ["time"]=>
      float(1451758047.7217)
    }
  }
  ["methodFoo"]=>
  array(1) {
    [0]=>
    array(3) {
      ["args"]=>
      array(0) {
      }
      ["result"]=>
      bool(true)
      ["time"]=>
      float(1451758047.7217)
    }
  }
}
```

**Comment by alexbowers** [comment]
Is this a bug for php internals then?

Perhaps a change to how __call works?

**Comment by davidpersson** [comment]
We need to investigate first. It might just be to the internals of Mocker.

**Comment by davidpersson** [comment]
Fixed by 593b315
